cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -170,69 +164,69 @@ protected void postProcess() {
+      return Collections.emptyList();
+    }
+    ProgramRunId programRunId = GSON.fromJson(programRun, ProgramRunId.class);
+
+    ProgramRunStatus programRunStatus = null;
+    if (programStatusStr != null) {
+      try {
+        programRunStatus = ProgramRunStatus.valueOf(programStatusStr);
+      } catch (IllegalArgumentException e) {
+        LOG.warn("Ignore notification with invalid program run status {} for program {}, {}",
+                 programStatusStr, programRun, notification);
         return Collections.emptyList();
       }
-      ProgramRunId programRunId = GSON.fromJson(programRun, ProgramRunId.class);
-
-      ProgramRunStatus programRunStatus = null;
-      if (programStatusStr != null) {
-        try {
-          programRunStatus = ProgramRunStatus.valueOf(programStatusStr);
-        } catch (IllegalArgumentException e) {
-          LOG.warn("Ignore notification with invalid program run status {} for program {}, {}",
-                   programStatusStr, programRun, notification);
-          return Collections.emptyList();
-        }
-      }
+    }
 
-      ProgramRunClusterStatus clusterStatus = null;
-      if (clusterStatusStr != null) {
-        try {
-          clusterStatus = ProgramRunClusterStatus.valueOf(clusterStatusStr);
-        } catch (IllegalArgumentException e) {
-          LOG.warn("Ignore notification with invalid program run cluster status {} for program {}",
-                   clusterStatusStr, programRun);
-          return Collections.emptyList();
-        }
-      }
-      if (notification.getNotificationType().equals(Notification.Type.PROGRAM_HEART_BEAT)) {
-        RunRecordMeta runRecordMeta = appMetadataStore.getRun(programRunId);
-        long heartBeatTimeInSeconds =
-          TimeUnit.MILLISECONDS.toSeconds(Long.parseLong(properties.get(ProgramOptionConstants.HEART_BEAT_TIME)));
-        writeToHeartBeatTable(runRecordMeta, heartBeatTimeInSeconds, programHeartbeatTable);
-        // we can return after writing to heart beat table
+    ProgramRunClusterStatus clusterStatus = null;
+    if (clusterStatusStr != null) {
+      try {
+        clusterStatus = ProgramRunClusterStatus.valueOf(clusterStatusStr);
+      } catch (IllegalArgumentException e) {
+        LOG.warn("Ignore notification with invalid program run cluster status {} for program {}",
+                 clusterStatusStr, programRun);
         return Collections.emptyList();
       }
-      List<Runnable> result = new ArrayList<>();
-      if (programRunStatus != null) {
-        handleProgramEvent(programRunId, programRunStatus, notification, messageIdBytes,
-                           appMetadataStore, programHeartbeatTable).ifPresent(result::add);
-      }
-      if (clusterStatus == null) {
-        return result;
-      }
+    }
+    if (notification.getNotificationType().equals(Notification.Type.PROGRAM_HEART_BEAT)) {
+      RunRecordMeta runRecordMeta = appMetadataStore.getRun(programRunId);
```

Comment:
doesn't this start another transaction?

Reply:
It will not since all the methods in `AppMetadataStore` do not start its own transaction.

Reply:
oh, it is just misnamed. Normally, dataset == no tx, store == dataset + tx. Got it